Sources/SwiftSyntaxMacrosTestSupport/Assertions.swift

```diff
@@ -218,70 +280,132 @@ extension DiagnosticSpec {
+  }
 
   if let highlights = spec.highlights {
     if diag.highlights.count != highlights.count {
-      XCTFail(
+      failureHandler(
         """
         Expected \(highlights.count) highlights but received \(diag.highlights.count):
         \(diag.highlights.map(\.trimmedDescription).joined(separator: "\n"))
         """,
-        file: spec.originatorFile,
-        line: spec.originatorLine
+        spec.originatorFileID,
+        spec.originatorFilePath,
+        spec.originatorLine,
+        spec.originatorColumn
       )
     } else {
-      for (actual, expected) in zip(diag.highlights, highlights) {
-        assertStringsEqualWithDiff(
-          actual.trimmedDescription,
-          expected,
-          "highlight does not match",
-          file: spec.originatorFile,
-          line: spec.originatorLine
-        )
+      for (actual, expected) in zip(diag.highlights.lazy.map(\.trimmedDescription), highlights) {
+        if actual != expected {
+          let message = describeDifferenceBetweenStrings(actual, expected, "highlight does not match")
+          failureHandler(
+            message,
+            spec.originatorFileID,
+            spec.originatorFilePath,
+            spec.originatorLine,
+            spec.originatorColumn
+          )
+        }
       }
     }
   }
   if diag.notes.count != spec.notes.count {
-    XCTFail(
+    failureHandler(
       """
       Expected \(spec.notes.count) notes but received \(diag.notes.count):
       \(diag.notes.map(\.debugDescription).joined(separator: "\n"))
       """,
-      file: spec.originatorFile,
-      line: spec.originatorLine
+      spec.originatorFileID,
+      spec.originatorFilePath,
+      spec.originatorLine,
+      spec.originatorColumn
     )
   } else {
     for (note, expectedNote) in zip(diag.notes, spec.notes) {
-      assertNote(note, in: expansionContext, expected: expectedNote)
+      assertNote(note, in: expansionContext, expected: expectedNote, failingWith: failureHandler)
     }
   }
   if diag.fixIts.count != spec.fixIts.count {
-    XCTFail(
+    failureHandler(
       """
       Expected \(spec.fixIts.count) Fix-Its but received \(diag.fixIts.count):
       \(diag.fixIts.map(\.message.message).joined(separator: "\n"))
       """,
-      file: spec.originatorFile,
-      line: spec.originatorLine
+      spec.originatorFileID,
+      spec.originatorFilePath,
+      spec.originatorLine,
+      spec.originatorColumn
     )
   } else {
     for (fixIt, expectedFixIt) in zip(diag.fixIts, spec.fixIts) {
-      assertFixIt(fixIt, expected: expectedFixIt)
+      assertFixIt(fixIt, expected: expectedFixIt, failingWith: failureHandler)
     }
   }
 }
 
+private func defaultFailureHandler(
+  _ message: String,
+  fileID: String = #fileID,
+  filePath: StaticString = #filePath,
+  line: UInt = #line,
+  column: UInt = #column
+) {
+  // FIXME: detect if XCTest is currently running a test and only call XCTFail() if it is.
+  // FIXME: support swift-testing.
```

Comment:
I'm not a huge fan of FIXME's that don't have corresponding issues. I'd just leave this out and file an issue to support swift-testing by default when we can.